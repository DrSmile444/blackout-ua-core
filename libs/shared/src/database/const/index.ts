export * from './user-location.const';
