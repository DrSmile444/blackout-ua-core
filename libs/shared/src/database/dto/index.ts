export * from './outrage.dto';
export * from './user.dto';
