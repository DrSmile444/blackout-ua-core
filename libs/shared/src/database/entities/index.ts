import { Outrage } from './outrage.entity';
import { OutrageQueue } from './outrage-queue.entity';
import { OutrageShift } from './outrage-shift.entity';
import { User } from './user.entity';
import { UserLocation } from './user-location.entity';

export * from './outrage.entity';
export * from './outrage.enums';
export * from './outrage-queue.entity';
export * from './outrage-shift.entity';
export * from './user.entity';
export * from './user-location.entity';

const userEntities = [User, UserLocation];
const outrageEntities = [Outrage, OutrageShift, OutrageQueue];

export const databaseEntities = [...userEntities, ...outrageEntities];
