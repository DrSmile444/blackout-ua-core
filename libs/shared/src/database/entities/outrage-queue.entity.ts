// outrage-queue.entity.ts
import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from 'typeorm';

import { LightStatus } from './outrage.enums';
// eslint-disable-next-line import/no-cycle
import { OutrageShift } from './outrage-shift.entity';

@Entity('outrage_queues')
export class OutrageQueue {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  queue: string;

  @Column({
    type: 'enum',
    enum: LightStatus,
  })
  lightStatus: LightStatus;

  @ManyToOne(() => OutrageShift, (shift) => shift.queues, { onDelete: 'CASCADE' })
  shift: OutrageShift;
}
