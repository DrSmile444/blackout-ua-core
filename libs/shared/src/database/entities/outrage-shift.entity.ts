// outrage-shift.entity.ts
import { Column, Entity, ManyToOne, OneToMany, PrimaryGeneratedColumn } from 'typeorm';

// eslint-disable-next-line import/no-cycle
import { Outrage } from './outrage.entity';
// eslint-disable-next-line import/no-cycle
import { OutrageQueue } from './outrage-queue.entity';

@Entity('outrage_shifts')
export class OutrageShift {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ type: 'timestamp' })
  start: Date;

  @Column({ type: 'timestamp' })
  end: Date;

  @ManyToOne(() => Outrage, (outrage) => outrage.shifts, { onDelete: 'CASCADE' })
  outrage: Outrage;

  @OneToMany(() => OutrageQueue, (queue) => queue.shift, { cascade: true, onDelete: 'CASCADE' })
  queues: OutrageQueue[];
}
