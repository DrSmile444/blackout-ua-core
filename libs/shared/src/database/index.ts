export * from './database.module';
export * from './dto';
export * from './entities';
export * from './services';
