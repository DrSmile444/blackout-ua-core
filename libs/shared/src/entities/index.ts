export * from './blackout-location.entity';
export * from './outrage.entity';
export * from './user.entity';
