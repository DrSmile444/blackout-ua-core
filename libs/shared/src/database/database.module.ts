import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { TypeOrmModule } from '@nestjs/typeorm';

import { BlackoutLocation, User } from '../entities';

import { UserService } from './services';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
    }),
    TypeOrmModule.forRootAsync({
      imports: [ConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'postgres',
        host: configService.get<string>('DATABASE_HOST'),
        port: configService.get<number>('DATABASE_PORT'),
        username: configService.get<string>('DATABASE_USERNAME'),
        password: configService.get<string>('DATABASE_PASSWORD'),
        database: configService.get<string>('DATABASE_NAME'),
        synchronize: true,
      }),
      inject: [ConfigService],
    }),
    TypeOrmModule.forFeature([User, BlackoutLocation]),
  ],
  providers: [UserService],
  exports: [UserService],
})
export class DatabaseModule {}
