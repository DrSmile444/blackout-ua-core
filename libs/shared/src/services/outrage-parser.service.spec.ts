import type { TestingModule } from '@nestjs/testing';
import { Test } from '@nestjs/testing';

import { OutrageRegion } from '../database/entities';
import { outrageMock5Change2, outrageMock6Origin, outrageMock7ChangeOld } from '../mocks';

import { OutrageParserService } from './outrage-parser.service';

describe('OutrageParserService', () => {
  let outrageParserService: OutrageParserService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [OutrageParserService],
    }).compile();

    outrageParserService = module.get(OutrageParserService);
  });

  describe('parseTime', () => {
    it('should return null if time period is not found', () => {
      const result = outrageParserService.parseTime('Some random string');
      expect(result).toBeNull();
    });

    it('should return null if start time is not found', () => {
      const result = outrageParserService.parseTime('Some random string 10:00–');
      expect(result).toBeNull();
    });

    it('should return null if end time is not found', () => {
      const result = outrageParserService.parseTime('Some random string –10:00');
      expect(result).toBeNull();
    });

    it('should return time period', () => {
      const result = outrageParserService.parseTime('Some random string 10:00–12:00');
      expect(result).toEqual(['10:00', '12:00']);
    });

    it('should return time period with spaces', () => {
      const result = outrageParserService.parseTime('Some random string 10:00 – 12:00');
      expect(result).toEqual(['10:00', '12:00']);
    });

    it('should return time period with dash', () => {
      const result = outrageParserService.parseTime('Some random string 10:00 - 12:00');
      expect(result).toEqual(['10:00', '12:00']);
    });

    it('should return time period with dash and spaces', () => {
      const result = outrageParserService.parseTime('Some random string 10:00 - 12:00');
      expect(result).toEqual(['10:00', '12:00']);
    });

    it('should return time period from real new mock', () => {
      const result = outrageParserService.parseTime(outrageMock6Origin);
      expect(result).toEqual(['00:00', '01:00']);
    });

    it('should return time period from real old mock', () => {
      const result = outrageParserService.parseTime(outrageMock7ChangeOld);
      expect(result).toEqual(['14:00', '15:00']);
    });

    // TODO add this logic
    // it('should return time period with dash and spaces and end time without minutes', () => {
    //   const result = outrageParserService.parseTime('Some random string 10:00 - 12');
    //   expect(result).toEqual(['10:00', '12']);
    // });
  });

  describe('parseQueue', () => {
    it('should return empty array if time period is not found', () => {
      const result = outrageParserService.parseQueue('Some random string');
      expect(result).toEqual([]);
    });

    it('should return empty array if time period is found but no queues', () => {
      const result = outrageParserService.parseQueue('Some random string 10:00–12:00');
      expect(result).toEqual([]);
    });

    it('should return queues', () => {
      const result = outrageParserService.parseQueue('Some random string 10:00–12:00 1, 2, 3');
      expect(result).toEqual(['1', '2', '3']);
    });

    it('should return 2 digit queues', () => {
      const result = outrageParserService.parseQueue('Some random string 10:00–12:00 1, 2, 13');
      expect(result).toEqual(['1', '2', '13']);
    });

    it('should return queues from real new mock', () => {
      const result = outrageParserService.parseQueue('01:00-02:00 1 та 2 черги');
      expect(result).toEqual(['1', '2']);
    });

    it('should return queues from real old mock', () => {
      const result = outrageParserService.parseQueue('➡️ 17:00 – 18:00 – 3');
      expect(result).toEqual(['3']);
    });
  });

  describe('parseRow', () => {
    it('should return null if time period is not found', () => {
      const result = outrageParserService.parseRow('Some random string');
      expect(result).toBeNull();
    });

    it('should return null if time period is found but no queues', () => {
      const result = outrageParserService.parseRow('Some random string 10:00–12:00');
      expect(result).toBeNull();
    });

    it('should return shift if time period and queues are found', () => {
      const result = outrageParserService.parseRow('Some random string 10:00–12:00 1, 2, 3');
      expect(result).toEqual({
        start: '10:00',
        end: '12:00',
        queues: ['1', '2', '3'],
      });
    });

    it('should return shift if time period and two digit queues are found', () => {
      const result = outrageParserService.parseRow('Some random string 10:00–12:00 1, 2, 13');
      expect(result).toEqual({
        start: '10:00',
        end: '12:00',
        queues: ['1', '2', '13'],
      });
    });

    it('should return shift from real new mock', () => {
      const result = outrageParserService.parseRow('01:00-02:00 1 та 2 черги');
      expect(result).toEqual({
        start: '01:00',
        end: '02:00',
        queues: ['1', '2'],
      });
    });

    it('should return shift from real new mock with multiple queues', () => {
      const result = outrageParserService.parseRow('22:00-23:00  2, 3 та 4 черги');
      expect(result).toEqual({
        start: '22:00',
        end: '23:00',
        queues: ['2', '3', '4'],
      });
    });

    it('should return shift from real old mock', () => {
      const result = outrageParserService.parseRow('➡️ 17:00 – 18:00 – 3');
      expect(result).toEqual({
        start: '17:00',
        end: '18:00',
        queues: ['3'],
      });
    });
  });

  describe('parseShifts', () => {
    it('should return empty array if message is empty', () => {
      const result = outrageParserService.parseShifts('');
      expect(result).toEqual([]);
    });

    it('should return empty array if message is not empty but no shifts', () => {
      const result = outrageParserService.parseShifts('Some random string');
      expect(result).toEqual([]);
    });

    it('should return shifts', () => {
      const result = outrageParserService.parseShifts('Some random string\n10:00–12:00 1, 2, 3\nSome random string');
      expect(result).toEqual([
        {
          start: '10:00',
          end: '12:00',
          queues: ['1', '2', '3'],
        },
      ]);
    });

    it('should return shifts from real new mock', () => {
      const result = outrageParserService.parseShifts(outrageMock6Origin);
      expect(result).toMatchSnapshot();
    });

    it('should return shifts from real old mock', () => {
      const result = outrageParserService.parseShifts(outrageMock7ChangeOld);
      expect(result).toMatchSnapshot();
    });
  });

  describe('parseDate', () => {
    it('should correctly parse the date from the message', () => {
      const message = 'відключень на 15 лютого';
      const result = outrageParserService.parseDate(message);
      const expectedDate = new Date(new Date().getFullYear(), 1, 15);
      expect(result).toEqual(expectedDate);
    });

    it('should correctly parse the date from real message', () => {
      const result = outrageParserService.parseDate(outrageMock5Change2);
      const expectedDate = new Date(new Date().getFullYear(), 5, 24);
      expect(result).toEqual(expectedDate);
    });

    it('should correctly parse the date from real message 2', () => {
      const result = outrageParserService.parseDate(outrageMock7ChangeOld);
      const expectedDate = new Date(new Date().getFullYear(), 5, 11);
      expect(result).toEqual(expectedDate);
    });

    it('should return null if the date is not found in the message', () => {
      const message = 'Some random string';
      const result = outrageParserService.parseDate(message);
      expect(result).toBeNull();
    });
  });

  describe('parseMessage', () => {
    it('should return parsed message', () => {
      const result = outrageParserService.parseMessage(outrageMock6Origin, OutrageRegion.CHERKASY);
      expect(result).toMatchSnapshot();
    });

    it('should return parsed message from real old mock', () => {
      const result = outrageParserService.parseMessage(outrageMock7ChangeOld, OutrageRegion.CHERKASY);
      expect(result).toMatchSnapshot();
    });
  });
});
