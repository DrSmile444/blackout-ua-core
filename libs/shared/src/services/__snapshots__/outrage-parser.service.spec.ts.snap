// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutrageParserService parseMessage should return parsed message 1`] = `
{
  "date": 2024-06-24T21:00:00.000Z,
  "region": "cherkasy",
  "shifts": [
    {
      "end": "01:00",
      "queues": [
        "3",
        "4",
      ],
      "start": "00:00",
    },
    {
      "end": "02:00",
      "queues": [
        "3",
        "4",
      ],
      "start": "01:00",
    },
    {
      "end": "03:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "02:00",
    },
    {
      "end": "04:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "03:00",
    },
    {
      "end": "05:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "04:00",
    },
    {
      "end": "06:00",
      "queues": [
        "2",
        "3",
      ],
      "start": "05:00",
    },
    {
      "end": "07:00",
      "queues": [
        "5",
        "6",
      ],
      "start": "06:00",
    },
    {
      "end": "08:00",
      "queues": [
        "5",
        "6",
      ],
      "start": "07:00",
    },
    {
      "end": "09:00",
      "queues": [
        "1",
        "2",
      ],
      "start": "08:00",
    },
    {
      "end": "10:00",
      "queues": [
        "1",
        "2",
      ],
      "start": "09:00",
    },
    {
      "end": "11:00",
      "queues": [
        "3",
        "4",
      ],
      "start": "10:00",
    },
    {
      "end": "12:00",
      "queues": [
        "3",
        "4",
      ],
      "start": "11:00",
    },
    {
      "end": "13:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "12:00",
    },
    {
      "end": "14:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "13:00",
    },
    {
      "end": "15:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "14:00",
    },
    {
      "end": "16:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "15:00",
    },
    {
      "end": "17:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "16:00",
    },
    {
      "end": "18:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "17:00",
    },
    {
      "end": "19:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "18:00",
    },
    {
      "end": "20:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "19:00",
    },
    {
      "end": "21:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "20:00",
    },
    {
      "end": "22:00",
      "queues": [
        "5",
        "6",
        "1",
      ],
      "start": "21:00",
    },
    {
      "end": "23:00",
      "queues": [
        "2",
        "3",
        "4",
      ],
      "start": "22:00",
    },
    {
      "end": "24:00",
      "queues": [
        "3",
        "4",
      ],
      "start": "23:00",
    },
  ],
  "type": "schedule",
}
`;

exports[`OutrageParserService parseMessage should return parsed message from real old mock 1`] = `
{
  "date": 2024-06-10T21:00:00.000Z,
  "region": "cherkasy",
  "shifts": [
    {
      "end": "15:00",
      "queues": [
        "2",
      ],
      "start": "14:00",
    },
    {
      "end": "16:00",
      "queues": [
        "2",
      ],
      "start": "15:00",
    },
    {
      "end": "17:00",
      "queues": [
        "3",
      ],
      "start": "16:00",
    },
    {
      "end": "18:00",
      "queues": [
        "3",
      ],
      "start": "17:00",
    },
    {
      "end": "19:00",
      "queues": [
        "4",
      ],
      "start": "18:00",
    },
    {
      "end": "20:00",
      "queues": [
        "4",
      ],
      "start": "19:00",
    },
    {
      "end": "21:00",
      "queues": [
        "5",
      ],
      "start": "20:00",
    },
    {
      "end": "22:00",
      "queues": [
        "5",
      ],
      "start": "21:00",
    },
    {
      "end": "23:00",
      "queues": [
        "6",
      ],
      "start": "22:00",
    },
  ],
  "type": "change",
}
`;

exports[`OutrageParserService parseShifts should return shifts from real new mock 1`] = `
[
  {
    "end": "01:00",
    "queues": [
      "3",
      "4",
    ],
    "start": "00:00",
  },
  {
    "end": "02:00",
    "queues": [
      "3",
      "4",
    ],
    "start": "01:00",
  },
  {
    "end": "03:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "02:00",
  },
  {
    "end": "04:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "03:00",
  },
  {
    "end": "05:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "04:00",
  },
  {
    "end": "06:00",
    "queues": [
      "2",
      "3",
    ],
    "start": "05:00",
  },
  {
    "end": "07:00",
    "queues": [
      "5",
      "6",
    ],
    "start": "06:00",
  },
  {
    "end": "08:00",
    "queues": [
      "5",
      "6",
    ],
    "start": "07:00",
  },
  {
    "end": "09:00",
    "queues": [
      "1",
      "2",
    ],
    "start": "08:00",
  },
  {
    "end": "10:00",
    "queues": [
      "1",
      "2",
    ],
    "start": "09:00",
  },
  {
    "end": "11:00",
    "queues": [
      "3",
      "4",
    ],
    "start": "10:00",
  },
  {
    "end": "12:00",
    "queues": [
      "3",
      "4",
    ],
    "start": "11:00",
  },
  {
    "end": "13:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "12:00",
  },
  {
    "end": "14:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "13:00",
  },
  {
    "end": "15:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "14:00",
  },
  {
    "end": "16:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "15:00",
  },
  {
    "end": "17:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "16:00",
  },
  {
    "end": "18:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "17:00",
  },
  {
    "end": "19:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "18:00",
  },
  {
    "end": "20:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "19:00",
  },
  {
    "end": "21:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "20:00",
  },
  {
    "end": "22:00",
    "queues": [
      "5",
      "6",
      "1",
    ],
    "start": "21:00",
  },
  {
    "end": "23:00",
    "queues": [
      "2",
      "3",
      "4",
    ],
    "start": "22:00",
  },
  {
    "end": "24:00",
    "queues": [
      "3",
      "4",
    ],
    "start": "23:00",
  },
]
`;

exports[`OutrageParserService parseShifts should return shifts from real old mock 1`] = `
[
  {
    "end": "15:00",
    "queues": [
      "2",
    ],
    "start": "14:00",
  },
  {
    "end": "16:00",
    "queues": [
      "2",
    ],
    "start": "15:00",
  },
  {
    "end": "17:00",
    "queues": [
      "3",
    ],
    "start": "16:00",
  },
  {
    "end": "18:00",
    "queues": [
      "3",
    ],
    "start": "17:00",
  },
  {
    "end": "19:00",
    "queues": [
      "4",
    ],
    "start": "18:00",
  },
  {
    "end": "20:00",
    "queues": [
      "4",
    ],
    "start": "19:00",
  },
  {
    "end": "21:00",
    "queues": [
      "5",
    ],
    "start": "20:00",
  },
  {
    "end": "22:00",
    "queues": [
      "5",
    ],
    "start": "21:00",
  },
  {
    "end": "23:00",
    "queues": [
      "6",
    ],
    "start": "22:00",
  },
]
`;
